// @dicy/types/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_OPTIONS = { $BIBINPUTS: ['$ROOTDIR',
        '$ROOTDIR/$OUTDIR',
        ''],
    $TEXINPUTS: ['$ROOTDIR',
        '$ROOTDIR/$OUTDIR',
        ''],
    bibtexEngine: 'bibtex',
    cleanPatterns: ['$OUTDIR/$JOB!($OUTEXT|.synctex.gz|.tex)',
        '/$OUTDIR/_minted-$JOB/*'],
    dviToPdfEngine: 'xdvipdfmx',
    engine: 'pdflatex',
    epstopdfBoundingBox: 'default',
    epstopdfOutputPath: '$DIR_0/$NAME_0.pdf',
    indexAutomaticRanges: true,
    indexEngine: 'makeindex',
    indexOrdering: 'word',
    indexSorting: 'default',
    knitrConcordance: true,
    knitrOutputPath: '$JOB.tex',
    lhs2texStyle: 'poly',
    literateAgdaEngine: 'agda',
    literateHaskellEngine: 'lhs2TeX',
    loadCache: true,
    loadUserOptions: true,
    outputFormat: 'pdf',
    phaseCycles: 20,
    pweaveCacheDirectory: 'pweave-cache-for-$JOB',
    pweaveFigureDirectory: 'pweave-figures-for-$JOB',
    pweaveKernel: 'python3',
    pweaveOutputFormat: 'tex',
    pweaveOutputPath: '$JOB.tex',
    saveCache: true,
    severity: 'warning',
    validateCache: true };
// END_AUTO
