// @dicy/core/lib/Rules/DviToPs.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("../Rule");
class DviToPs extends Rule_1.default {
    static isApplicable(consumer, command, phase, parameters = []) {
        return __awaiter(this, void 0, void 0, function* () {
            // Only apply if output format is ps or intermediate PostScript generation
            // is on.
            return consumer.options.outputFormat === 'ps' ||
                (consumer.options.outputFormat === 'pdf' && !!consumer.options.intermediatePostScript);
        });
    }
    constructCommand() {
        return {
            args: [
                'dvips',
                '-o',
                '{{$DIR_0/$NAME_0.ps}}',
                '{{$FILEPATH_0}}'
            ],
            cd: '$ROOTDIR',
            severity: 'error',
            inputs: [{ file: '$FILEPATH_0', type: 'target' }],
            outputs: [{ file: '$DIR_0/$NAME_0.ps', type: 'target' }]
        };
    }
}
DviToPs.parameterTypes = [new Set(['DeviceIndependentFile'])];
DviToPs.description = 'Converts DVI to PS using dvips.';
exports.default = DviToPs;
