// @dicy/core/lib/Rules/DviToSvg.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("../Rule");
class DviToSvg extends Rule_1.default {
    static isApplicable(consumer, command, phase, parameters = []) {
        return __awaiter(this, void 0, void 0, function* () {
            // Only apply if output format is svg
            return consumer.options.outputFormat === 'svg';
        });
    }
    constructCommand() {
        return {
            args: [
                'dvisvgm',
                '-o',
                '{{$DIR_0/$NAME_0.svg}}',
                '{{$FILEPATH_0}}'
            ],
            cd: '$ROOTDIR',
            severity: 'error',
            inputs: [{ file: '$FILEPATH_0', type: 'target' }],
            outputs: [{ file: '$DIR_0/$NAME_0.svg', type: 'target' }]
        };
    }
}
DviToSvg.parameterTypes = [new Set(['DeviceIndependentFile'])];
DviToSvg.description = 'Converts DVI to SVG using dvisvgm.';
exports.default = DviToSvg;
