// @dicy/core/lib/Rules/MetaPost.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("../Rule");
class MetaPost extends Rule_1.default {
    getFileActions(file) {
        return __awaiter(this, void 0, void 0, function* () {
            // ParsedFileListing triggers updateDependencies, all others trigger run.
            return [file.type === 'ParsedFileListing' ? 'updateDependencies' : 'run'];
        });
    }
    constructCommand() {
        // Force the same error options as LaTeX and capture the file listing.
        return {
            args: [
                'mpost',
                '-file-line-error',
                '-interaction=batchmode',
                '-recorder',
                '{{$BASE_0}}'
            ],
            cd: '$ROOTDIR_0',
            severity: 'error',
            inputs: [{ file: '$DIR_0/$NAME_0.fls-ParsedFileListing' }],
            outputs: [
                { file: '$DIR_0/$NAME_0.fls' },
                { file: '$DIR_0/$NAME_0.log' }
            ]
        };
    }
}
MetaPost.parameterTypes = [new Set(['MetaPost'])];
MetaPost.description = 'Runs MetaPost on produced MetaPost files.';
exports.default = MetaPost;
