// @dicy/core/lib/Rules/ParseXindyLog.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Rule_1 = require("../Rule");
class ParsedXindyLog extends Rule_1.default {
    /**
     * Parse the xindy log.
     * @return {Promise<boolean>}  Status of rule evaluation
     */
    parse() {
        return __awaiter(this, void 0, void 0, function* () {
            // Get the output file
            const output = yield this.getResolvedOutput('$FILEPATH_0-ParsedXindyLog');
            if (!output)
                return false;
            const parsedLog = {
                messages: [],
                inputs: [],
                outputs: [],
                calls: []
            };
            const name = 'xindy';
            let filePath;
            yield this.firstParameter.parse([{
                    // Error/Warning messages with file references
                    names: ['severity', 'text', 'file'],
                    patterns: [/^(ERROR|WARNING): (.*? in:?)$/i, /(.*)/],
                    evaluate: (mode, reference, match) => {
                        parsedLog.messages.push({
                            name,
                            severity: match.groups.severity.toLowerCase(),
                            text: `${match.groups.text} ${match.groups.file}`,
                            source: { file: path.normalize(match.groups.file) },
                            log: reference
                        });
                    }
                }, {
                    // Error/Warning messages
                    names: ['severity', 'text'],
                    patterns: [/^(ERROR|WARNING): (.*)$/i],
                    evaluate: (mode, reference, match) => {
                        parsedLog.messages.push({
                            name,
                            severity: match.groups.severity.toLowerCase(),
                            text: match.groups.text,
                            source: { file: filePath },
                            log: reference
                        });
                    }
                }, {
                    // Module loadings
                    names: ['text'],
                    patterns: [/^(Loading module .*|Finished loading module .*)$/i],
                    evaluate: (mode, reference, match) => {
                        parsedLog.messages.push({
                            name,
                            severity: 'info',
                            text: match.groups.text,
                            source: { file: filePath },
                            log: reference
                        });
                    }
                }, {
                    // Input files
                    names: ['file'],
                    patterns: [/^Reading raw-index (.*?)[.]{3}$/i],
                    evaluate: (mode, reference, match) => {
                        filePath = path.normalize(match.groups.file);
                        parsedLog.inputs.push(path.normalize(match.groups.file));
                        parsedLog.messages.push({
                            name,
                            severity: 'info',
                            text: match._,
                            source: { file: filePath },
                            log: reference
                        });
                    }
                }, {
                    // Output files
                    names: ['file'],
                    patterns: [/^(?:Markup written into file|Opening logfile) (.*?)[. ]$/i],
                    evaluate: (mode, reference, match) => {
                        parsedLog.outputs.push(path.normalize(match.groups.file));
                        parsedLog.messages.push({
                            name,
                            severity: 'info',
                            text: match._,
                            source: { file: filePath },
                            log: reference
                        });
                    }
                }]);
            output.value = parsedLog;
            return true;
        });
    }
}
ParsedXindyLog.parameterTypes = [new Set(['XindyLog'])];
ParsedXindyLog.commands = new Set(['build', 'log']);
ParsedXindyLog.description = 'Parses the logs produced by xindy and texindy.';
ParsedXindyLog.defaultActions = ['parse'];
exports.default = ParsedXindyLog;
