// @dicy/core/lib/Rules/ParseLaTeXAuxilary.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("../Rule");
class ParseLaTeXAuxilary extends Rule_1.default {
    parse() {
        return __awaiter(this, void 0, void 0, function* () {
            const output = yield this.getResolvedOutput('$FILEPATH_0-ParsedLaTeXAuxilary');
            if (!output)
                return false;
            const results = {
                commands: []
            };
            // We just need a list of the commands that in the aux file so we know
            // when to run BibTeX and Bib2Gls.
            yield this.firstParameter.parse([{
                    names: ['command'],
                    patterns: [/^\\([A-Za-z@]+)/],
                    evaluate: (mode, reference, match) => {
                        results.commands.push(match.groups.command);
                    }
                }]);
            output.value = results;
            return true;
        });
    }
}
ParseLaTeXAuxilary.parameterTypes = [new Set(['LaTeXAuxilary'])];
ParseLaTeXAuxilary.defaultActions = ['parse'];
ParseLaTeXAuxilary.description = 'Parses the aux files produced by all variants of LaTeX.';
exports.default = ParseLaTeXAuxilary;
