// @dicy/core/lib/Rules/ParseBibToGlsLog.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("../Rule");
class ParseBibTeXLog extends Rule_1.default {
    parse() {
        return __awaiter(this, void 0, void 0, function* () {
            const output = yield this.getResolvedOutput('$FILEPATH_0-ParsedBibToGlsLog');
            if (!output)
                return false;
            const parsedLog = {
                messages: [],
                inputs: [],
                outputs: [],
                calls: []
            };
            yield this.firstParameter.parse([{
                    names: ['output'],
                    patterns: [/^Writing (.*)$/],
                    evaluate: (mode, reference, match) => {
                        parsedLog.outputs.push(this.normalizePath(match.groups.output));
                    }
                }, {
                    names: ['input'],
                    patterns: [/^Reading (.*)$/],
                    evaluate: (mode, reference, match) => {
                        parsedLog.inputs.push(this.normalizePath(match.groups.input));
                    }
                }]);
            output.value = parsedLog;
            return true;
        });
    }
}
ParseBibTeXLog.parameterTypes = [new Set(['BibToGlsLog'])];
ParseBibTeXLog.commands = new Set(['build', 'log']);
ParseBibTeXLog.defaultActions = ['parse'];
ParseBibTeXLog.description = 'Parses any bib2gls produced logs.';
exports.default = ParseBibTeXLog;
