// @dicy/core/lib/Rules/ParseAsymptoteStdOut.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Rule_1 = require("../Rule");
class ParseAsymptoteStdOut extends Rule_1.default {
    parse() {
        return __awaiter(this, void 0, void 0, function* () {
            const output = yield this.getResolvedOutput('$DIR_0/$NAME_0.log-ParsedAsymptoteStdOut');
            if (!output)
                return false;
            let rootPath = this.rootPath;
            const parsedLog = {
                messages: [],
                inputs: [],
                outputs: [],
                calls: []
            };
            yield this.firstParameter.parse([{
                    names: ['filePath'],
                    patterns: [/^cd (.*)$/],
                    evaluate: (mode, reference, match) => {
                        rootPath = match.groups.filePath;
                    }
                }, {
                    names: ['filePath'],
                    patterns: [/^Wrote (.*)$/],
                    evaluate: (mode, reference, match) => {
                        parsedLog.outputs.push(this.normalizePath(path.resolve(rootPath, match.groups.filePath)));
                    }
                }, {
                    names: ['type', 'filePath'],
                    patterns: [/^(Including|Loading) \S+ from (.*)$/],
                    evaluate: (mode, reference, match) => {
                        if (!path.isAbsolute(match.groups.filePath)) {
                            parsedLog.inputs.push(this.normalizePath(path.resolve(rootPath, match.groups.filePath)));
                        }
                    }
                }]);
            output.value = parsedLog;
            return true;
        });
    }
}
ParseAsymptoteStdOut.parameterTypes = [new Set(['AsymptoteStdOut'])];
ParseAsymptoteStdOut.defaultActions = ['parse'];
ParseAsymptoteStdOut.description = 'Parses the console output of Asymptote.';
exports.default = ParseAsymptoteStdOut;
