// @dicy/core/lib/Rules/ParseSplitIndexStdOut.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Rule_1 = require("../Rule");
class ParseSplitIndexStdOut extends Rule_1.default {
    parse() {
        return __awaiter(this, void 0, void 0, function* () {
            const output = yield this.getResolvedOutput('$DIR_0/$NAME_0.log-ParsedSplitIndexStdOut');
            if (!output)
                return false;
            const parsedLog = {
                messages: [],
                inputs: [],
                outputs: [],
                calls: []
            };
            yield this.firstParameter.parse([{
                    names: ['text'],
                    patterns: [/^(Close.*|New index file.*)$/],
                    evaluate: (mode, reference, match) => {
                        // Do not include the reference since it is to a virtual file.
                        parsedLog.messages.push({
                            severity: 'info',
                            name: 'splitindex',
                            text: match.groups.text
                        });
                    }
                }, {
                    names: ['filePath'],
                    patterns: [/^(.*?) with [0-9]+ lines$/],
                    evaluate: (mode, reference, match) => {
                        // Do not include the reference since it is to a virtual file.
                        parsedLog.messages.push({
                            severity: 'info',
                            name: 'splitindex',
                            text: match._
                        });
                        parsedLog.outputs.push(this.normalizePath(path.resolve(this.rootPath, match.groups.filePath)));
                    }
                }]);
            output.value = parsedLog;
            return true;
        });
    }
}
ParseSplitIndexStdOut.parameterTypes = [new Set(['SplitIndexStdOut'])];
ParseSplitIndexStdOut.defaultActions = ['parse'];
ParseSplitIndexStdOut.description = 'Parses the console output of splitindex.';
exports.default = ParseSplitIndexStdOut;
