// @dicy/core/lib/Rules/PsToPdf.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("../Rule");
class PsToPdf extends Rule_1.default {
    static isApplicable(consumer, command, phase, parameters = []) {
        return __awaiter(this, void 0, void 0, function* () {
            // Only apply if output format is pdf
            return consumer.options.outputFormat === 'pdf';
        });
    }
    constructCommand() {
        return {
            args: [
                'ps2pdf',
                '{{$FILEPATH_0}}',
                '{{$DIR_0/$NAME_0.pdf}}'
            ],
            cd: '$ROOTDIR',
            severity: 'error',
            inputs: [{ file: '$FILEPATH_0', type: 'target' }],
            outputs: [{ file: '$DIR_0/$NAME_0.pdf', type: 'target' }]
        };
    }
}
PsToPdf.parameterTypes = [new Set(['PostScript'])];
PsToPdf.description = 'Converts PS to PDF using ps2pdf.';
exports.default = PsToPdf;
